### ISSUE TITLE
Glob handling is too brittle

### ISSUE THREAD

**Issue by McDutchie** [att/ast#1021]
The script below reliably makes ksh hang. (mksh, bash, yash, and ash variants execute it in no time at all.)

```
case '\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\' in
( *\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\*\\* )
	echo ok ;;
esac

```

**Comment by krader1961** [comment]
@McDutchie Can you provide more context? That is, a real world example? What prompted you to notice this issue? Because I would tell anyone who wrote that code to pound sand :smile:

**Comment by krader1961** [comment]
Also, I just ran your test case from my home directory using ksh93u, bash, and zsh on my macOS platform and both "hang". That is, they are all cpu bound and spend huge amounts of time expanding those patterns. So when you say the non-ksh variants "execute it in no time at all" my experience is quite different.

**Comment by McDutchie** [comment]
Yes, zsh hangs, and I just reported it to them too. In fact I found it there first.

bash really doesn't, though -- neither 3.2.57 nor 5.0beta. Try again.

Use case is irrelevant. Brittle code is bad, and indicative of other potential problems. This shouldn't hang.

(Sorry for brevity, I'm very short on time and sleep atm.)

**Comment by krader1961** [comment]
It doesn't hang. It simply takes several minutes to expand the glob. On my system from my home directory which includes symlinks to other filesystems that are huge it takes ~5 minutes for each shell. Which is not surprising since the glob is pathological. This [fish issue](https://github.com/fish-shell/fish-shell/issues/5240) may be relevant.

The behavior you see, @McDutchie, compared to what I see is probably due to differences in the specific file system namespace between your system and mine. For example, the presence of symlinks to other file systems.

I don't doubt that there are pathological situations such as your example which could, and should, be handled more efficiently. But this seems like a low priority issue relative to all the other problems with the ksh code. At least unless you can provide a more realistic example.

**Comment by McDutchie** [comment]
Eh? The file system doesn't enter into it. This is glob string processing with `case`, not pathname expansion. No I/O happens at any time.

The fact that it takes five minutes to process a single string is far more pathological than that glob. It's evidence of brokenness.

**Comment by krader1961** [comment]
@McDutchie You're right, of course. I was thinking of how fish handles globs in a switch statement. I also meant to link to https://github.com/fish-shell/fish-shell/issues/4094 which in turn links to this [paper](https://research.swtch.com/glob) about the poor performance of most shells and showing an algorithm that is considerably more efficient. And while your example doesn't involve pathname expansion I'm pretty sure the same glob evaluation logic is used for both cases. And thus implementing the solution in that paper would fix the performance of your pathological example.
